Source/Core/DiscIO/VolumeVerifier.cpp

```diff
@@ -0,0 +1,648 @@
+                "the disc image as several parts, you need to merge them into one file.");
+    AddProblem(Severity::High, text);
+    return;
+  }
+
+  if (ShouldBeDualLayer() && biggest_offset <= SL_DVD_R_SIZE)
+  {
+    AddProblem(
+        Severity::Medium,
+        GetStringT("This game has been hacked to fit on a single-layer DVD. Some content such as "
+                   "pre-rendered videos, extra languages or entire game modes will be broken. "
+                   "This problem generally only exists in illegal copies of games."));
+  }
+
+  if (!m_volume.IsSizeAccurate())
+  {
+    AddProblem(Severity::Low, GetStringT("The format that the disc image is saved in does not "
+                                         "store the size of the disc image."));
+  }
+  else if (!m_is_tgc)
+  {
+    const Platform platform = m_volume.GetVolumeType();
+    const u64 size = m_volume.GetSize();
+
+    const bool valid_gamecube = size == MINI_DVD_SIZE;
+    const bool valid_retail_wii = size == SL_DVD_SIZE || size == DL_DVD_SIZE;
+    const bool valid_debug_wii = size == SL_DVD_R_SIZE || size == DL_DVD_R_SIZE;
+
+    const bool debug = IsDebugSigned();
+    if ((platform == Platform::GameCubeDisc && !valid_gamecube) ||
+        (platform == Platform::WiiDisc && (debug ? !valid_debug_wii : !valid_retail_wii)))
+    {
+      if (debug && valid_retail_wii)
+      {
+        AddProblem(Severity::Low,
+                   GetStringT("This debug disc image has the size of a retail disc image."));
+      }
+      else
+      {
+        const bool small =
+            (m_volume.GetVolumeType() == Platform::GameCubeDisc && size < MINI_DVD_SIZE) ||
+            (m_volume.GetVolumeType() == Platform::WiiDisc && size < SL_DVD_SIZE);
+
+        if (small)
+        {
+          AddProblem(Severity::Low,
+                     GetStringT("This disc image has an unusual size. This will likely make the "
+                                "emulated loading times longer. When using netplay or sending "
+                                "input recordings to other people, you will likely experience "
+                                "desyncs if anyone is using a good dump."));
+        }
+        else
+        {
+          AddProblem(Severity::Low, GetStringT("This disc image has an unusual size."));
+        }
+      }
+    }
+  }
+}
+
+u64 VolumeVerifier::GetBiggestUsedOffset()
+{
+  std::vector<Partition> partitions = m_volume.GetPartitions();
+  if (partitions.empty())
+    partitions.emplace_back(m_volume.GetGamePartition());
+
+  u64 biggest_offset = 0;
+  for (const Partition& partition : partitions)
+  {
+    if (partition != PARTITION_NONE)
+      biggest_offset = std::max(biggest_offset, partition.offset);
+
+    const std::optional<u64> fst_offset = GetFSTOffset(m_volume, partition);
+    const std::optional<u64> fst_size = GetFSTSize(m_volume, partition);
+    if (!fst_offset || !fst_size)
+      break;
```

Comment:
Should this really be a `break` and not just a `continue`? Same for the one below.

Reply:
Oops, you're right.